# Regression Problem on PISA

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt # this is for visualization
import seaborn as sns # for visualization
%matplotlib inline
import statsmodels.formula.api as sm
import scipy, scipy.stats
import math # log function is inside this library
# this is used to get the plots inline i.e. in the same page
from collections import Counter

In [5]:
df1 = pd.read_csv("pisa2009train.csv")
df2 = pd.read_csv("pisa2009test.csv")

In [6]:
df1.head()

grade  male   raceeth  preschool  expectBachelors  motherHS  \
0     11     1       NaN        NaN              0.0       NaN   
1     11     1     White        0.0              0.0       1.0   
2      9     1     White        1.0              1.0       1.0   
3     10     0     Black        1.0              1.0       0.0   
4     10     1  Hispanic        1.0              0.0       1.0   

   motherBachelors  motherWork  fatherHS  fatherBachelors      ...       \
0              NaN         1.0       NaN              NaN      ...        
1              1.0         1.0       1.0              0.0      ...        
2              1.0         1.0       1.0              NaN      ...        
3              0.0         1.0       1.0              0.0      ...        
4              0.0         1.0       1.0              0.0      ...        

   englishAtHome  computerForSchoolwork  read30MinsADay  \
0            0.0                    1.0             0.0   
1            1.0                    1.0             1.0   
2            1.0                    1.0             0.0   
3            1.0                    1.0             1.0   
4            1.0                    1.0             1.0   

   minutesPerWeekEnglish  studentsInEnglish  schoolHasLibrary  publicSchool  \
0                  225.0                NaN               1.0             1   
1                  450.0               25.0               1.0             1   
2                  250.0               28.0               1.0             1   
3                  200.0               23.0               1.0             1   
4                  250.0               35.0               1.0             1   

   urban  schoolSize  readingScore  
0      1       673.0        476.00  
1      0      1173.0        575.01  
2      0      1233.0        554.81  
3      1      2640.0        458.11  
4      1      1095.0        613.89  

[5 rows x 24 columns]

In [7]:
df2.head()

grade  male raceeth  preschool  expectBachelors  motherHS  motherBachelors  \
0     10     0   White        1.0              0.0       1.0              1.0   
1     10     1   White        0.0              0.0       1.0              0.0   
2     10     0   White        1.0              0.0       1.0              0.0   
3     10     0   White        1.0              0.0       1.0              1.0   
4     10     0   White        1.0              1.0       1.0              0.0   

   motherWork  fatherHS  fatherBachelors      ...       englishAtHome  \
0         1.0       1.0              0.0      ...                 1.0   
1         1.0       1.0              0.0      ...                 1.0   
2         1.0       1.0              0.0      ...                 1.0   
3         1.0       1.0              0.0      ...                 1.0   
4         0.0       1.0              1.0      ...                 1.0   

   computerForSchoolwork  read30MinsADay  minutesPerWeekEnglish  \
0                    1.0             0.0                  240.0   
1                    1.0             0.0                  255.0   
2                    1.0             0.0                    NaN   
3                    1.0             0.0                  160.0   
4                    1.0             0.0                  240.0   

   studentsInEnglish  schoolHasLibrary  publicSchool  urban  schoolSize  \
0               30.0               1.0             1      0       808.0   
1                NaN               1.0             1      0       808.0   
2               30.0               1.0             1      0       808.0   
3               30.0               NaN             1      0       808.0   
4               30.0               1.0             1      0       808.0   

   readingScore  
0        355.24  
1        385.57  
2        522.62  
3        406.24  
4        453.50  

[5 rows x 24 columns]

In [8]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3663 entries, 0 to 3662
Data columns (total 24 columns):
grade                    3663 non-null int64
male                     3663 non-null int64
raceeth                  3628 non-null object
preschool                3607 non-null float64
expectBachelors          3601 non-null float64
motherHS                 3566 non-null float64
motherBachelors          3266 non-null float64
motherWork               3570 non-null float64
fatherHS                 3418 non-null float64
fatherBachelors          3094 non-null float64
fatherWork               3430 non-null float64
selfBornUS               3594 non-null float64
motherBornUS             3592 non-null float64
fatherBornUS             3550 non-null float64
englishAtHome            3592 non-null float64
computerForSchoolwork    3598 non-null float64
read30MinsADay           3629 non-null float64
minutesPerWeekEnglish    3477 non-null float64
studentsInEnglish        3414 non-null float64
schoo

Ques1. Load the training and testing sets using the read_csv() function, and save them as dataframes with the names pisaTrain and pisaTest. 
How many students are there in the training set?

Ans. Total Number of students in the training set are 3663.

Ques2. Using Group by on pisaTrain, what is the average reading test score of males?
Of Females?


Ans. First Count the No. Males and Females

In [9]:
df1['male'].value_counts()

1    1872
0    1791
Name: male, dtype: int64

The Number of Males are 1872 while Number of Females are 1791

In [10]:
df1.groupby('male')['readingScore'].agg([np.mean])

mean
male            
0     512.940631
1     483.532479

Average reading test score of males is 483.53 while Average reading test score of Female is 512.94

Ques3. Which variables have missing data in at least one observation in the training set? 

In [11]:
df1.isnull().sum()

grade                      0
male                       0
raceeth                   35
preschool                 56
expectBachelors           62
motherHS                  97
motherBachelors          397
motherWork                93
fatherHS                 245
fatherBachelors          569
fatherWork               233
selfBornUS                69
motherBornUS              71
fatherBornUS             113
englishAtHome             71
computerForSchoolwork     65
read30MinsADay            34
minutesPerWeekEnglish    186
studentsInEnglish        249
schoolHasLibrary         143
publicSchool               0
urban                      0
schoolSize               162
readingScore               0
dtype: int64

Ans. The coulmns with missing value in the traing set are raceteeth, preschool, expectBachelors,motherHS, motherBachelors, motherWork,fatherHS, fatherBachelors, fatherWork, selfBornUS, motherBornUS, englishAtHome, computerForSchoolWork, read30MinsADay, minutesPerWeekEnglish,studentsInEnglish,schoolHasLibrary & schoolSize.        

Ques 4. Linear regression discards observations with missing data, so we will remove all such observations from the training and testing sets. Ideally we should be filling the missing values with plausible information, but for now, lets remove them and move ahead. How many observations are now in the training set? How many observations are now in the testing set?

In [12]:
df1 = df1.dropna(axis=0)

In [13]:
df2.head()

grade  male raceeth  preschool  expectBachelors  motherHS  motherBachelors  \
0     10     0   White        1.0              0.0       1.0              1.0   
1     10     1   White        0.0              0.0       1.0              0.0   
2     10     0   White        1.0              0.0       1.0              0.0   
3     10     0   White        1.0              0.0       1.0              1.0   
4     10     0   White        1.0              1.0       1.0              0.0   

   motherWork  fatherHS  fatherBachelors      ...       englishAtHome  \
0         1.0       1.0              0.0      ...                 1.0   
1         1.0       1.0              0.0      ...                 1.0   
2         1.0       1.0              0.0      ...                 1.0   
3         1.0       1.0              0.0      ...                 1.0   
4         0.0       1.0              1.0      ...                 1.0   

   computerForSchoolwork  read30MinsADay  minutesPerWeekEnglish  \
0                    1.0             0.0                  240.0   
1                    1.0             0.0                  255.0   
2                    1.0             0.0                    NaN   
3                    1.0             0.0                  160.0   
4                    1.0             0.0                  240.0   

   studentsInEnglish  schoolHasLibrary  publicSchool  urban  schoolSize  \
0               30.0               1.0             1      0       808.0   
1                NaN               1.0             1      0       808.0   
2               30.0               1.0             1      0       808.0   
3               30.0               NaN             1      0       808.0   
4               30.0               1.0             1      0       808.0   

   readingScore  
0        355.24  
1        385.57  
2        522.62  
3        406.24  
4        453.50  

[5 rows x 24 columns]

In [14]:
df2 = df2.dropna(axis=0)

In [15]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2414 entries, 1 to 3662
Data columns (total 24 columns):
grade                    2414 non-null int64
male                     2414 non-null int64
raceeth                  2414 non-null object
preschool                2414 non-null float64
expectBachelors          2414 non-null float64
motherHS                 2414 non-null float64
motherBachelors          2414 non-null float64
motherWork               2414 non-null float64
fatherHS                 2414 non-null float64
fatherBachelors          2414 non-null float64
fatherWork               2414 non-null float64
selfBornUS               2414 non-null float64
motherBornUS             2414 non-null float64
fatherBornUS             2414 non-null float64
englishAtHome            2414 non-null float64
computerForSchoolwork    2414 non-null float64
read30MinsADay           2414 non-null float64
minutesPerWeekEnglish    2414 non-null float64
studentsInEnglish        2414 non-null float64
schoo

In [16]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 990 entries, 0 to 1569
Data columns (total 24 columns):
grade                    990 non-null int64
male                     990 non-null int64
raceeth                  990 non-null object
preschool                990 non-null float64
expectBachelors          990 non-null float64
motherHS                 990 non-null float64
motherBachelors          990 non-null float64
motherWork               990 non-null float64
fatherHS                 990 non-null float64
fatherBachelors          990 non-null float64
fatherWork               990 non-null float64
selfBornUS               990 non-null float64
motherBornUS             990 non-null float64
fatherBornUS             990 non-null float64
englishAtHome            990 non-null float64
computerForSchoolwork    990 non-null float64
read30MinsADay           990 non-null float64
minutesPerWeekEnglish    990 non-null float64
studentsInEnglish        990 non-null float64
schoolHasLibrary         

Ans. The Number of observations in the training set are 2414 and The Number of observations in testing set are 990


Ques5.	Which of the following variables is an unordered factor/categorical variable with at least 3 levels? 

In [17]:
df1['raceeth'].value_counts()

White                                     1470
Hispanic                                   500
Black                                      228
Asian                                       95
More than one race                          81
American Indian/Alaska Native               20
Native Hawaiian/Other Pacific Islander      20
Name: raceeth, dtype: int64

In [18]:
df2['raceeth'].value_counts()

White                                     613
Hispanic                                  201
Black                                      79
Asian                                      45
More than one race                         35
American Indian/Alaska Native              11
Native Hawaiian/Other Pacific Islander      6
Name: raceeth, dtype: int64

Ans. Raceeth is the variable which is an unodered factor/categorical variable with at least 3 levels. 

Ques6. If we select White as the reference level, which binary variables will be included in the regression model? (Select all that apply.)

Ans. Acc. to the condition if we select White as the reference level, we will include Hispanic, Black, Asian, More than one race, American Indian/Alaska Native, Native Hawaiin/Other Pacific Islander. 

Ques7. Consider again adding our unordered Category race to the regression model with reference level "White". 

For a student who is Asian, which binary variables would be set to 0? All remaining variables will be set to 1. 


Ans. For a student who is Asian, the binary variable which will be set to 0 are Hispanic, Black, More Than one race, American Indian/Alaska Native, Native Hawaiian/Other Pacific Islander. 

Ques8. For a student who is white, which binary variables would be set to 0? All remaining variables will be set to 1.

For a student who is white, the binary variables that would be set to 0 are Hispanic, Black, Asian, More than one race, American Indian/Alska Native, Native Hawaiian/Other Pacific Islander. 

Ques9. Because the race variable takes on text values, we will have to create Dummy Variables before we run Modeling on the data. 

Build a linear regression model (call it lmScore) using the training set to predict readingScore using all the remaining variables. (Including the dummies). Remember to omit one of the dummy variable ( the omitted dummy will be the reference) 

What is the Multiple R-squared value of lmScore on the training set?


In [19]:
df1 = pd.get_dummies(df1,drop_first=True)
df1.head()

grade  male  preschool  expectBachelors  motherHS  motherBachelors  \
1     11     1        0.0              0.0       1.0              1.0   
3     10     0        1.0              1.0       0.0              0.0   
4     10     1        1.0              0.0       1.0              0.0   
7     10     0        1.0              1.0       1.0              0.0   
9     10     1        1.0              1.0       1.0              1.0   

   motherWork  fatherHS  fatherBachelors  fatherWork      ...        \
1         1.0       1.0              0.0         1.0      ...         
3         1.0       1.0              0.0         1.0      ...         
4         1.0       1.0              0.0         0.0      ...         
7         0.0       1.0              0.0         1.0      ...         
9         1.0       0.0              0.0         1.0      ...         

   publicSchool  urban  schoolSize  readingScore  raceeth_Asian  \
1             1      0      1173.0        575.01              0   
3             1      1      2640.0        458.11              0   
4             1      1      1095.0        613.89              0   
7             1      0      1913.0        439.36              0   
9             1      0       899.0        465.90              0   

   raceeth_Black  raceeth_Hispanic  raceeth_More than one race  \
1              0                 0                           0   
3              1                 0                           0   
4              0                 1                           0   
7              0                 0                           0   
9              0                 0                           1   

   raceeth_Native Hawaiian/Other Pacific Islander  raceeth_White  
1                                               0              1  
3                                               0              0  
4                                               0              0  
7                                               0              1  
9                                               0              0  

[5 rows x 29 columns]

In [20]:
df1 = df1.drop(['raceeth_White'], axis=1)

In [21]:
df1.head()

grade  male  preschool  expectBachelors  motherHS  motherBachelors  \
1     11     1        0.0              0.0       1.0              1.0   
3     10     0        1.0              1.0       0.0              0.0   
4     10     1        1.0              0.0       1.0              0.0   
7     10     0        1.0              1.0       1.0              0.0   
9     10     1        1.0              1.0       1.0              1.0   

   motherWork  fatherHS  fatherBachelors  fatherWork  \
1         1.0       1.0              0.0         1.0   
3         1.0       1.0              0.0         1.0   
4         1.0       1.0              0.0         0.0   
7         0.0       1.0              0.0         1.0   
9         1.0       0.0              0.0         1.0   

                        ...                        schoolHasLibrary  \
1                       ...                                     1.0   
3                       ...                                     1.0   
4                       ...                                     1.0   
7                       ...                                     1.0   
9                       ...                                     1.0   

   publicSchool  urban  schoolSize  readingScore  raceeth_Asian  \
1             1      0      1173.0        575.01              0   
3             1      1      2640.0        458.11              0   
4             1      1      1095.0        613.89              0   
7             1      0      1913.0        439.36              0   
9             1      0       899.0        465.90              0   

   raceeth_Black  raceeth_Hispanic  raceeth_More than one race  \
1              0                 0                           0   
3              1                 0                           0   
4              0                 1                           0   
7              0                 0                           0   
9              0                 0                           1   

   raceeth_Native Hawaiian/Other Pacific Islander  
1                                               0  
3                                               0  
4                                               0  
7                                               0  
9                                               0  

[5 rows x 28 columns]

In [22]:
from sklearn.linear_model import LinearRegression

In [23]:
lmScore=LinearRegression()
lmScore.fit(df1.iloc[:,[0,1,2,3,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,23,24,25,26,27]],df1['readingScore'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [24]:
lmScore.intercept_

141.8838859613142

In [25]:
lmScore.coef_

array([ 2.95775015e+01, -1.48059544e+01, -4.42852098e+00,  5.54138008e+01,
        1.30911485e+01, -2.44033758e+00,  6.58506638e+00,  1.72113365e+01,
        5.76467329e+00, -4.38745087e+00, -8.26754712e+00,  4.21836521e+00,
        9.27649884e+00,  2.49986142e+01,  3.44407746e+01,  1.24756805e-02,
       -2.57063241e-01,  1.24532903e+01, -1.79175284e+01, -5.35500471e-01,
        6.88673799e-03, -3.44356841e+00, -6.56206658e+01, -3.79858170e+01,
       -1.57138989e+01, -4.59313444e+00])

In [26]:
df1['PredictScore']=lmScore.predict(df1.iloc[:,[0,1,2,3,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,23,24,25,26,27]])

In [27]:
TSS = sum((df1['readingScore']-np.mean(df1['readingScore']))**2)
SSE = sum((df1['readingScore']-df1['PredictScore'])**2)

In [29]:
R2= 1-(SSE/TSS)
print(R2)

0.3203690464849498


Ans. R2 Value of lmScore on the training Set is 0.32

Ques10. What is the training-set root-mean squared error (RMSE) of lmScore?

In [ ]:
n = len(df1['PredictScore'])
rmse = np.linalg.norm( df1['PredictScore']- df1['readingScore']) / np.sqrt(n)
rmse


Ques11.	Consider two students A and B. They have all variable values the same, except that student A is in grade 11 and student B is in grade 9. What is the predicted reading score of student A minus the predicted reading score of student B?

In [ ]:
df1.groupby('grade')['PredictScore'].agg([np.mean])

Ques12.	What is the meaning of the coefficient associated with variable raceethAsian?

The coefficient associated with the variable raceethAsian is -3.44356841e+00 which is the predicted diff between an Asian Student and American Indian/Alaska Native student

In [ ]:
df1.corr()

Ques13.	Based on the significance codes, which variables out of these are candidates for removal from the model? Select all that apply. (The factor variable raceeth should only be removed if none of its levels are significant.)

Ans. The raceeth should be removed because some of the variables has low value of the correlation with the reading scores.

Ques14. Using the "predict" function , use the lmScore model to predict the reading scores of students in pisaTest. Call this list of predictions "predTest". Do not change the variables in the model (for example, do not remove variables that we found were not significant in the previous part of this problem). 

What is the range between the maximum and minimum predicted reading score on the test set?


In [135]:
df2 = df2.dropna(axis=0)

In [ ]:
#Divide both the data same as Titanic. 

In [30]:
df2 = pd.get_dummies(df2,drop_first=True)
df2.head()

grade  male  preschool  expectBachelors  motherHS  motherBachelors  \
0     10     0        1.0              0.0       1.0              1.0   
4     10     0        1.0              1.0       1.0              0.0   
6     10     0        1.0              0.0       1.0              0.0   
7     10     0        1.0              0.0       1.0              0.0   
8     11     0        0.0              0.0       1.0              1.0   

   motherWork  fatherHS  fatherBachelors  fatherWork      ...        \
0         1.0       1.0              0.0         0.0      ...         
4         0.0       1.0              1.0         1.0      ...         
6         0.0       1.0              0.0         0.0      ...         
7         1.0       1.0              0.0         1.0      ...         
8         1.0       1.0              1.0         1.0      ...         

   publicSchool  urban  schoolSize  readingScore  raceeth_Asian  \
0             1      0       808.0        355.24              0   
4             1      0       808.0        453.50              0   
6             1      0       808.0        405.13              0   
7             1      0       808.0        665.05              0   
8             1      0       808.0        604.84              0   

   raceeth_Black  raceeth_Hispanic  raceeth_More than one race  \
0              0                 0                           0   
4              0                 0                           0   
6              0                 0                           0   
7              0                 0                           0   
8              0                 0                           0   

   raceeth_Native Hawaiian/Other Pacific Islander  raceeth_White  
0                                               0              1  
4                                               0              1  
6                                               0              0  
7                                               0              1  
8                                               0              1  

[5 rows x 29 columns]

In [31]:
df2 = df2.drop(['raceeth_White'], axis=1)

In [32]:
df2.head()

grade  male  preschool  expectBachelors  motherHS  motherBachelors  \
0     10     0        1.0              0.0       1.0              1.0   
4     10     0        1.0              1.0       1.0              0.0   
6     10     0        1.0              0.0       1.0              0.0   
7     10     0        1.0              0.0       1.0              0.0   
8     11     0        0.0              0.0       1.0              1.0   

   motherWork  fatherHS  fatherBachelors  fatherWork  \
0         1.0       1.0              0.0         0.0   
4         0.0       1.0              1.0         1.0   
6         0.0       1.0              0.0         0.0   
7         1.0       1.0              0.0         1.0   
8         1.0       1.0              1.0         1.0   

                        ...                        schoolHasLibrary  \
0                       ...                                     1.0   
4                       ...                                     1.0   
6                       ...                                     1.0   
7                       ...                                     1.0   
8                       ...                                     1.0   

   publicSchool  urban  schoolSize  readingScore  raceeth_Asian  \
0             1      0       808.0        355.24              0   
4             1      0       808.0        453.50              0   
6             1      0       808.0        405.13              0   
7             1      0       808.0        665.05              0   
8             1      0       808.0        604.84              0   

   raceeth_Black  raceeth_Hispanic  raceeth_More than one race  \
0              0                 0                           0   
4              0                 0                           0   
6              0                 0                           0   
7              0                 0                           0   
8              0                 0                           0   

   raceeth_Native Hawaiian/Other Pacific Islander  
0                                               0  
4                                               0  
6                                               0  
7                                               0  
8                                               0  

[5 rows x 28 columns]

In [37]:
X=df2.drop(labels=['readingScore'],axis=1)
Y=df2.readingScore

lmScore=LinearRegression()
lmScore.fit(X,Y)

df2['predTest']=lmScore.predict(X)
print('Difference between min and max test scores on testing set is',round(max(predTest)-min(predTest)))

Difference between min and max test scores on testing set is 282.0


In [39]:
SSE = sum((df2['readingScore']-df2['predTest'])**2)
SSE

5553662.813721299

In [140]:
lmScore2.intercept_

2.2737367544323206e-13

In [141]:
lmScore2.coef_

array([-1.89302394e-14, -1.50102153e-13,  1.14973451e-13,  8.34356731e-14,
        2.54496760e-14,  5.89356731e-15, -6.57637190e-15, -1.56004049e-14,
        5.22936050e-15,  7.05187072e-15,  3.35787766e-14, -5.18808498e-14,
        1.84388753e-15,  5.56358605e-15, -3.43273490e-15,  1.44605522e-14,
        5.42308320e-16,  8.67578578e-16, -6.77941539e-14, -3.83612277e-14,
       -1.19314579e-13,  5.55111512e-17,  1.00000000e+00,  1.90568861e-16,
       -4.36541807e-15, -1.00070099e-14, -5.78278287e-15])

In [142]:
df2['PredictScore1']=lmScore2.predict(df2.iloc[:,[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,24,25,26,27]])

In [143]:
df2['PredictScore1'].max()-df2['PredictScore1'].min()

529.8200000000002

Ques15.What is the sum of squared errors (SSE) of lmScore on the testing set? 
_____ (give a integer) 


In [145]:
SSE = sum((df2['readingScore']-df2['PredictScore1'])**2)
SSE

2.14889244679059e-23

In [45]:
inp = [{'c1':10, 'c2':100}, {'c1':11,'c2':110}, {'c1':12,'c2':120}]
df = pd.DataFrame(inp)
print (df)

   c1   c2
0  10  100
1  11  110
2  12  120


In [47]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from pandas import Series,DataFrame

import warnings
warnings.filterwarnings('ignore')

import scipy.stats as stats
import statsmodels.regression.linear_model as lmr
import statsmodels.discrete.discrete_model as sm
from statsmodels.api import add_constant

from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from collections import Counter
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.metrics import r2_score
from statsmodels.formula.api import ols

%matplotlib inline

In [63]:
for i, j in DataFrame(df.columns).iterrows():
    print(DataFrame(df.columns)[0][i])


c1
c2


In [64]:
df.columns

Index(['c1', 'c2'], dtype='object')